test/abort-controller.js

```diff
@@ -0,0 +1,204 @@
+'use strict'
+
+const { test } = require('tap')
+const { AbortController } = require('abort-controller')
+const { Client, errors } = require('..')
+const { createServer } = require('http')
+const { createReadStream } = require('fs')
+
+test('Abort before sending request (no body)', (t) => {
+  t.plan(3)
+
+  let count = 0
+  const server = createServer((req, res) => {
+    if (count === 1) {
+      t.fail('The second request should never be executed')
+    }
+    count += 1
+    res.end('hello')
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`)
+    const abortController = new AbortController()
+    t.teardown(client.destroy.bind(client))
+
+    client.request({ path: '/', method: 'GET' }, (err, response) => {
+      t.error(err)
+      const bufs = []
+      response.body.on('data', (buf) => {
+        bufs.push(buf)
+      })
+      response.body.on('end', () => {
+        t.strictEqual('hello', Buffer.concat(bufs).toString('utf8'))
+      })
+    })
+
+    client.request({ path: '/', method: 'GET', signal: abortController.signal }, (err, response) => {
+      t.ok(err instanceof errors.RequestAbortedError)
+    })
+
+    abortController.abort()
+  })
+})
+
+test('Abort while waiting response (no body)', (t) => {
+  t.plan(1)
+
+  const abortController = new AbortController()
+  const server = createServer((req, res) => {
+    abortController.abort()
+    res.setHeader('content-type', 'text/plain')
+    res.end('hello world')
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`)
+    t.teardown(client.destroy.bind(client))
+
+    client.request({ path: '/', method: 'GET', signal: abortController.signal }, (err, response) => {
+      t.ok(err instanceof errors.RequestAbortedError)
+    })
+  })
+})
+
+test('Abort while waiting response (write headers started) (no body)', (t) => {
+  t.plan(1)
+
+  const abortController = new AbortController()
+  const server = createServer((req, res) => {
+    res.writeHead(200, { 'content-type': 'text/plain' })
+    setTimeout(() => {
+      abortController.abort()
+      res.end('hello world')
+    })
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`)
+    t.teardown(client.destroy.bind(client))
+
+    client.request({ path: '/', method: 'GET', signal: abortController.signal }, (err, response) => {
+      t.ok(err instanceof errors.RequestAbortedError)
+    })
+  })
+})
```

Comment:
This test is a little weird? Shouldn't the callback be invoked with `writeHead`? i.e. this test should fail with `t.ok(err)`.

Reply:
`writeHead` doesn't flush.